### ISSUE TITLE
source and weight input channels mismatch when using Conv2D on mps

### ISSUE THREAD

**Issue by mcgibbon** [pytorch/pytorch#84206]
### 🐛 Describe the bug

When using 2D convolution on an M1 macbook with the mps device, python crashes. This can be reproduced with:

```python
import numpy as np
import torch
device = torch.device("mps")

conv = torch.nn.Conv2d(in_channels=1, out_channels=1, kernel_size=3).to(device)

data = torch.tensor(np.random.uniform(size=(1, 10, 10, 1)), dtype=torch.float32).to(device)
optimizer = torch.optim.Adam(conv.parameters(), lr=0.1)
x = data.permute(0, 3, 1, 2)
out = torch.sum(conv(x))
loss = torch.nn.MSELoss()(out, torch.zeros_like(out))
optimizer.zero_grad()
out.backward()
optimizer.step()
```

Which on my machine outputs:
```bash
$ python3 reproducer.py
/AppleInternal/Library/BuildRoots/560148d7-a559-11ec-8c96-4add460b61a6/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Runtimes/MPSRuntime/Operations/GPUConv2DOps.mm:214: failed assertion `Source and weight input channels mismatch'
zsh: abort      python3 reproducer.py
```

If I run the example above with the "cpu" device, it exits without error.

### Versions

```bash
$ python3 collect_env.py
Collecting environment information...
PyTorch version: 1.13.0.dev20220826
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.3.1 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2)
CMake version: version 3.23.1
Libc version: N/A

Python version: 3.8.13 | packaged by conda-forge | (default, Mar 25 2022, 06:04:14)  [Clang 12.0.1 ] (64-bit runtime)
Python platform: macOS-12.3.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.21.5
[pip3] torch==1.13.0.dev20220826
[pip3] torchaudio==0.13.0.dev20220826
[pip3] torchvision==0.14.0.dev20220826
[conda] numpy                     1.21.5           py38hb29071a_0    conda-forge
[conda] torch                     1.13.0.dev20220826          pypi_0    pypi
[conda] torchaudio                0.13.0.dev20220826          pypi_0    pypi
[conda] torchvision               0.14.0.dev20220826          pypi_0    pypi
```

cc @kulinseth @albanD

**Comment by zhaoBowen612** [comment]
meet same problems

**Comment by mcgibbon** [comment]
The permutation is necessary for the bug to show up. Using a tensor that already has channels-first at initialization and doesn't need a permute works just fine.

**Comment by mcgibbon** [comment]
OK, the following example does execute without error, so `.to(memory_format=torch.contiguous_format)` can be used as a workaround for now.

```
import numpy as np
import torch

device = torch.device("mps")

conv = torch.nn.Conv2d(in_channels=1, out_channels=1, kernel_size=3).to(device)

data = torch.tensor(np.random.uniform(size=(1, 10, 10, 1)), dtype=torch.float32).to(device)
optimizer = torch.optim.Adam(conv.parameters(), lr=0.1)
x = data.permute(0, 3, 1, 2).to(memory_format=torch.contiguous_format)
out = torch.sum(conv(x))
loss = torch.nn.MSELoss()(out, torch.zeros_like(out))
optimizer.zero_grad()
out.backward()
optimizer.step()
```

**Comment by kulinseth** [comment]
@mcgibbon , this issue is fixed in nightly. I tried locally and it seems to work. Can you please check?

**Comment by mcgibbon** [comment]
Confirmed, works for me too. THanks!

**Comment by Birch-san** [comment]
btw I'm still experiencing this error message on `1.14.0.dev20221013`, [`1.13.0` release candidate](https://download.pytorch.org/whl/test/cpu/torch-1.13.0-cp310-none-macosx_11_0_arm64.whl) and `1.12.1` stable.

not via the reproducer here, admittedly. so certainly you've fixed something there. it's hitting me via a different route.

when I run huggingface diffusers' CLIP-guided stable-diffusion example, it crashes on the backward pass:  
https://github.com/Birch-san/stable-diffusion/blob/3c3008fa7aa1d1c6da441db1fa82c9eae35b65db/scripts/diffusers_play.py#L164

likewise, CLIP-guided diffusion crashes on my CompVis-style stable-diffusion script, but **only** if I enable classifier-free guidance:  
https://github.com/Birch-san/stable-diffusion/blob/30eb6109156f8c6fbb694789ba3f6bb4867ca1fc/scripts/txt2img_fork.py#L166

error message is once again:

```
/AppleInternal/Library/BuildRoots/20d6c351-ee94-11ec-bcaf-7247572f23b4/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Runtimes/MPSRuntime/Operations/GPUConv2DOps.mm:214: failed assertion `Source and weight input channels mismatch'
```

I tried making contiguous all the tensors that I passed into the model, but unfortunately that didn't fix it.

still: there's got to be a clue there in the "only crashes when CFG guidance is used".

are there any techniques I can use to narrow this down?

**Comment by jdack41** [comment]
@kulinseth 
Any update for this?
I encountered same issue on diffusers fine tuning.
It seems this happen when backward with grad that has device info.
I also tried [this](https://zhuanlan.zhihu.com/p/555372346?utm_id=0)  workaround 
`loss = torch.Tensor([loss.item()]).requires_grad_()`
but this doesn’t backward grad properly.

**Comment by KonstantinRoehrl** [comment]
I also still face this issue on a Macbook Pro with M1 Pro chip. Installed torch versions have been 1.14.0.dev*** and now the new 2.0.0.dev***.

The error message is as follows:
`AppleInternal/Library/BuildRoots/d7fdb87a-93e6-11ec-b907-4a23b0182bfd/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Runtimes/MPSRuntime/Operations/GPUConv2DOps.mm:214: failed assertion 'Source and weight input channels mismatch'`

which apparently leads to (not sure, if this is relevant to torch but included it for completeness):
`/Users/konstantin/opt/anaconda3/envs/sporeagent/lib/python3.9/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '`

@mcgibbon mentioned that the permutation is necessary for the bug to show up and a possible workaround would be to use `.to(memory_format=torch.contiguous_format)`. Unfortunately, i can not confirm this, because there are no calls to permute(...) in my code and i still get this error. Therefore, i can't use any workaround. If i run my code on the cpu, everything works as expected.

If i figure out a minimal reproducible example, i will post it here.
